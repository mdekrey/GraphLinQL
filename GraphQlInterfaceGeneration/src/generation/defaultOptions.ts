import { Options } from "./Options";

export const defaultOptions: Options = {
  namespace: "GraphLinqQL.Interfaces",
  useNullabilityIndicator: true,
  scalarTypes: {
    Int: { csharpNullable: false, csharpType: "int" },
    Float: { csharpNullable: false, csharpType: "double" },
    String: { csharpNullable: true, csharpType: "string" },
    Boolean: { csharpNullable: false, csharpType: "bool" }
  },
  using: [
    "GraphLinqQL",
    "GraphLinqQL.Introspection",
    "System",
    "System.Collections",
    "System.Collections.Generic",
    "System.Collections.Immutable",
    "System.Linq"
  ],
  deserializer: "System.Text.Json.JsonSerializer.Deserialize",
  introspection: false
};
