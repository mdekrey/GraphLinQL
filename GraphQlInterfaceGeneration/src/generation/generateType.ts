import { GraphQLObjectType, GraphQLField, isNonNullType, GraphQLArgument } from "graphql";
import { Options } from "./Options";
import { getTypeName } from "./getTypeName";
import { getPropertyName } from "./getPropertyName";
import { getFieldName } from "./getFieldName";
import { getOutputTypeName } from "./getOutputTypeName";
import { getInputTypeName } from "./getInputTypeName";
import { toCsharpValue } from "./toCsharpValue";

export function generateType(object: GraphQLObjectType, options: Options) {
  const typeName = getTypeName(object.name, options);

  const fields = object.getFields();
  return `${
    object.description
      ? `
/// <summary>
/// ${object.description.split("\n").join("\n/// ")}
/// </summary>`
      : `
`
  }
public abstract class ${typeName} : IGraphQlResolvable${interfaceDeclaration(object, options)}
{
    private ${typeName}() { }
    ${Object.keys(fields)
      .map(fieldName => fields[fieldName])
      .map(field => resultAbstractDeclaration(field, options)).join(`
    `)}

    IGraphQlResult IGraphQlResolvable.ResolveQuery(string name, IGraphQlParameterResolver parameters)
    {
        switch (name)
        {
            case "__typename": return GraphQlConstantResult.Construct("${object.name}");
            ${Object.keys(fields)
              .map(fieldName => fields[fieldName])
              .map(field => fieldResolveQueryCase(field, options)).join(`
            `)}
            default: throw new ArgumentException("Unknown property " + name, "name");
        };
    }

    bool IGraphQlResolvable.IsType(string value)
    {
        return ${[object.name]
          .concat(object.getInterfaces().map(iface => iface.name))
          .map(n => `value == "${n}"`)
          .join(" || ")};
    }

    public abstract class GraphQlContract<T> : ${typeName}, IGraphQlAccepts<T>
    {
        public IGraphQlResultFactory<T> Original { get; set; }
        IGraphQlResultFactory IGraphQlAccepts.Original { set { Original = (IGraphQlResultFactory<T>)value; } }
        Type IGraphQlAccepts.ModelType { get { return typeof(T); } }
    }
}
`;
}

function interfaceDeclaration(object: GraphQLObjectType, options: Options) {
  if (object.getInterfaces().length) {
    return (
      ", " +
      object
        .getInterfaces()
        .map(iface => getTypeName(iface.name, options))
        .join(", ")
    );
  }
  return "";
}

function resultAbstractDeclaration(field: GraphQLField<any, any, { [key: string]: any }>, options: Options) {
  const propertyName = getPropertyName(field.name, options);
  const typeName = getOutputTypeName(field.type, options);
  const args = field.args.map(arg => resultAbstractDeclarationArg(arg, options)).join(", ");
  const obsoleteAttr = field.isDeprecated
    ? `[Obsolete(${field.deprecationReason ? JSON.stringify(field.deprecationReason) : ""})]
    `
    : ``;
  return `${
    field.description
      ? `
    /// <summary>
    /// ${field.description.split("\n").join("\n    /// ")}
    /// </summary>
    `
      : ``
  }${obsoleteAttr}public abstract IGraphQlResult${typeName && `<${typeName}>`} ${propertyName}(${args});`;
}

function resultAbstractDeclarationArg(arg: GraphQLArgument, options: Options) {
  const fieldName = getFieldName(arg.name, options);
  const inputTypeName = getInputTypeName(arg.type, options);
  return `${inputTypeName} ${fieldName}`;
}

function fieldResolveQueryCase(field: GraphQLField<any, any, { [key: string]: any }>, options: Options) {
  const propertyName = getPropertyName(field.name, options);
  const args = field.args.map(arg => fieldResolveQueryCaseArg(arg, options));
  return `case "${field.name}": return this.${propertyName}(${
    args.length
      ? `
                ${args.join(`,
                `)}`
      : ""
  });`;
}

function fieldResolveQueryCaseArg(arg: GraphQLArgument, options: Options) {
  const fieldName = getFieldName(arg.name, options);
  const inputTypeName = getInputTypeName(arg.type, options);
  const nullable = arg.defaultValue || !isNonNullType(arg.type);
  const defaultValue = arg.defaultValue;

  const getValue = nullable
    ? `(parameters.HasParameter("${arg.name}") ? parameters.GetParameter<${inputTypeName}>("${arg.name}") : null)`
    : `parameters.GetParameter<${inputTypeName}>("${arg.name}")`;
  const defaultValueExpression = defaultValue ? ` ?? ${toCsharpValue(arg.defaultValue, arg.type, options)}` : "";
  return `${fieldName}: ${getValue}${defaultValueExpression}`;
}
