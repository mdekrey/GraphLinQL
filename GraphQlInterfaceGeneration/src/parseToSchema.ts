import { parse, buildASTSchema, GraphQLSchema } from "graphql";

export function parseToSchema(graphQlSchema: string): GraphQLSchema {
  const astDocument = parse(graphQlSchema);

  const backcompatOptions = { commentDescriptions: true };
  const schema = buildASTSchema(astDocument, backcompatOptions);

  return schema;
}
