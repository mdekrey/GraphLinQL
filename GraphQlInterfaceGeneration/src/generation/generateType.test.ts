import { parseToSchema } from "../parseToSchema";
import { getStarWarsGraphQlSchemaV4 } from "../resources.test";
import { GraphQLObjectType } from "graphql";
import { defaultOptions } from "./defaultOptions";
import { generateType } from "./generateType";

it("can generate simple types", async function() {
  const schema = await getSampleSchema();
  const type = schema.getType("Film") as GraphQLObjectType;
  expect(generateType(type, defaultOptions)).toMatchSnapshot();
});

async function getSampleSchema() {
  const schema = parseToSchema(await getStarWarsGraphQlSchemaV4());
  return schema;
}
