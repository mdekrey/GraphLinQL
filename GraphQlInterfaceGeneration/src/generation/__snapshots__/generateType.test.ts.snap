// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can generate v4 types 1`] = `
"
/// <summary>
/// An individual person or character within the Star Wars universe.
/// </summary>
public abstract class Person : IGraphQlResolvable, Node
{
    private Person() { }
    
    /// <summary>
    /// The name of this person.
    /// </summary>
    public abstract IGraphQlResult<string?> name();
    
    /// <summary>
    /// The birth year of the person, using the in-universe standard of BBY or ABY -
    /// Before the Battle of Yavin or After the Battle of Yavin. The Battle of Yavin is
    /// a battle that occurs at the end of Star Wars episode IV: A New Hope.
    /// </summary>
    public abstract IGraphQlResult<string?> birthYear();
    
    /// <summary>
    /// The eye color of this person. Will be \\"unknown\\" if not known or \\"n/a\\" if the
    /// person does not have an eye.
    /// </summary>
    public abstract IGraphQlResult<string?> eyeColor();
    
    /// <summary>
    /// The gender of this person. Either \\"Male\\", \\"Female\\" or \\"unknown\\",
    /// \\"n/a\\" if the person does not have a gender.
    /// </summary>
    public abstract IGraphQlResult<string?> gender();
    
    /// <summary>
    /// The hair color of this person. Will be \\"unknown\\" if not known or \\"n/a\\" if the
    /// person does not have hair.
    /// </summary>
    public abstract IGraphQlResult<string?> hairColor();
    
    /// <summary>
    /// The height of the person in centimeters.
    /// </summary>
    public abstract IGraphQlResult<int?> height();
    
    /// <summary>
    /// The mass of the person in kilograms.
    /// </summary>
    public abstract IGraphQlResult<double?> mass();
    
    /// <summary>
    /// The skin color of this person.
    /// </summary>
    public abstract IGraphQlResult<string?> skinColor();
    
    /// <summary>
    /// A planet that this person was born on or inhabits.
    /// </summary>
    public abstract IGraphQlResult<Planet?> homeworld();
    public abstract IGraphQlResult<PersonFilmsConnection?> filmConnection(string? after, int? first, string? before, int? last);
    
    /// <summary>
    /// The species that this person belongs to, or null if unknown.
    /// </summary>
    public abstract IGraphQlResult<Species?> species();
    public abstract IGraphQlResult<PersonStarshipsConnection?> starshipConnection(string? after, int? first, string? before, int? last);
    public abstract IGraphQlResult<PersonVehiclesConnection?> vehicleConnection(string? after, int? first, string? before, int? last);
    
    /// <summary>
    /// The ISO 8601 date format of the time that this resource was created.
    /// </summary>
    public abstract IGraphQlResult<string?> created();
    
    /// <summary>
    /// The ISO 8601 date format of the time that this resource was edited.
    /// </summary>
    public abstract IGraphQlResult<string?> edited();
    
    /// <summary>
    /// The ID of an object
    /// </summary>
    public abstract IGraphQlResult<string> id();

    IGraphQlResult IGraphQlResolvable.ResolveQuery(string name, IGraphQlParameterResolver parameters) =>
        name switch
        {
            \\"__typename\\" => GraphQlConstantResult.Construct(\\"Person\\"),
            \\"name\\" => this.name(),
            \\"birthYear\\" => this.birthYear(),
            \\"eyeColor\\" => this.eyeColor(),
            \\"gender\\" => this.gender(),
            \\"hairColor\\" => this.hairColor(),
            \\"height\\" => this.height(),
            \\"mass\\" => this.mass(),
            \\"skinColor\\" => this.skinColor(),
            \\"homeworld\\" => this.homeworld(),
            \\"filmConnection\\" => this.filmConnection(
                after: (parameters.HasParameter(\\"after\\") ? parameters.GetParameter<string?>(\\"after\\") : null),
                first: (parameters.HasParameter(\\"first\\") ? parameters.GetParameter<int?>(\\"first\\") : null),
                before: (parameters.HasParameter(\\"before\\") ? parameters.GetParameter<string?>(\\"before\\") : null),
                last: (parameters.HasParameter(\\"last\\") ? parameters.GetParameter<int?>(\\"last\\") : null)),
            \\"species\\" => this.species(),
            \\"starshipConnection\\" => this.starshipConnection(
                after: (parameters.HasParameter(\\"after\\") ? parameters.GetParameter<string?>(\\"after\\") : null),
                first: (parameters.HasParameter(\\"first\\") ? parameters.GetParameter<int?>(\\"first\\") : null),
                before: (parameters.HasParameter(\\"before\\") ? parameters.GetParameter<string?>(\\"before\\") : null),
                last: (parameters.HasParameter(\\"last\\") ? parameters.GetParameter<int?>(\\"last\\") : null)),
            \\"vehicleConnection\\" => this.vehicleConnection(
                after: (parameters.HasParameter(\\"after\\") ? parameters.GetParameter<string?>(\\"after\\") : null),
                first: (parameters.HasParameter(\\"first\\") ? parameters.GetParameter<int?>(\\"first\\") : null),
                before: (parameters.HasParameter(\\"before\\") ? parameters.GetParameter<string?>(\\"before\\") : null),
                last: (parameters.HasParameter(\\"last\\") ? parameters.GetParameter<int?>(\\"last\\") : null)),
            \\"created\\" => this.created(),
            \\"edited\\" => this.edited(),
            \\"id\\" => this.id(),
            _ => throw new ArgumentException(\\"Unknown property \\" + name, nameof(name))
        };

    bool IGraphQlResolvable.IsType(string value) =>
      value == \\"Person\\" || value == \\"Node\\";

    public abstract class GraphQlContract<T> : Person, IGraphQlAccepts<T>
    {
#nullable disable
        public IGraphQlResultFactory<T> Original { get; set; }
#nullable enable

        IGraphQlResultFactory IGraphQlAccepts.Original { set { Original = (IGraphQlResultFactory<T>)value; } }
        Type IGraphQlAccepts.ModelType => typeof(T);
    }
}
"
`;

exports[`can generate various default values 1`] = `
"

public abstract class Query : IGraphQlResolvable
{
    private Query() { }
    public abstract IGraphQlResult<string?> randomTest(string? id, int? i, double? f, bool? b, IEnumerable<int>? l, string? s, ColorInput? o, ColorInput? n);

    IGraphQlResult IGraphQlResolvable.ResolveQuery(string name, IGraphQlParameterResolver parameters) =>
        name switch
        {
            \\"__typename\\" => GraphQlConstantResult.Construct(\\"Query\\"),
            \\"randomTest\\" => this.randomTest(
                id: (parameters.HasParameter(\\"id\\") ? parameters.GetParameter<string?>(\\"id\\") : null) ?? System.Text.Json.JsonSerializer.Deserialize<string?>(\\"\\\\\\"foo\\\\\\"\\"),
                i: (parameters.HasParameter(\\"i\\") ? parameters.GetParameter<int?>(\\"i\\") : null) ?? 15,
                f: (parameters.HasParameter(\\"f\\") ? parameters.GetParameter<double?>(\\"f\\") : null) ?? 2.5,
                b: (parameters.HasParameter(\\"b\\") ? parameters.GetParameter<bool?>(\\"b\\") : null) ?? true,
                l: (parameters.HasParameter(\\"l\\") ? parameters.GetParameter<IEnumerable<int>?>(\\"l\\") : null) ?? System.Text.Json.JsonSerializer.Deserialize<IEnumerable<int>?>(\\"[5,4,3]\\"),
                s: (parameters.HasParameter(\\"s\\") ? parameters.GetParameter<string?>(\\"s\\") : null) ?? \\"evil\\\\\\"\\\\\\\\\\\\\\"string\\\\\\\\\\",
                o: (parameters.HasParameter(\\"o\\") ? parameters.GetParameter<ColorInput?>(\\"o\\") : null) ?? System.Text.Json.JsonSerializer.Deserialize<ColorInput?>(\\"{\\\\\\"red\\\\\\":255,\\\\\\"green\\\\\\":255,\\\\\\"blue\\\\\\":255}\\"),
                n: (parameters.HasParameter(\\"n\\") ? parameters.GetParameter<ColorInput?>(\\"n\\") : null)),
            _ => throw new ArgumentException(\\"Unknown property \\" + name, nameof(name))
        };

    bool IGraphQlResolvable.IsType(string value) =>
      value == \\"Query\\";

    public abstract class GraphQlContract<T> : Query, IGraphQlAccepts<T>
    {
#nullable disable
        public IGraphQlResultFactory<T> Original { get; set; }
#nullable enable

        IGraphQlResultFactory IGraphQlAccepts.Original { set { Original = (IGraphQlResultFactory<T>)value; } }
        Type IGraphQlAccepts.ModelType => typeof(T);
    }
}
"
`;
